be/src/exprs/agg/combinator/state_merge_function.h

```diff
@@ -0,0 +1,124 @@
+                                                _agg_state_desc.get_return_type(), _agg_state_desc.get_arg_types());
+        return Status::OK();
+    }
+
+    StatusOr<ColumnPtr> execute(FunctionContext* context, const Columns& columns) override {
+        if (columns.size() != 1) {
+            return Status::InternalError("StateMergeFunction execute columns is not 1");
+        }
+        if (columns.size() != _arg_nullables.size()) {
+            return Status::InternalError("StateMergeFunction execute columns size " + std::to_string(columns.size()) +
+                                         " not match with arg_nullables size " + std::to_string(_arg_nullables.size()));
+        }
+
+        SCOPED_THREAD_LOCAL_AGG_STATE_ALLOCATOR_SETTER(&kDefaultAggStateMergeFunctionAllocator);
+
+        bool is_result_nullable = _agg_state_desc.is_result_nullable() || _arg_nullables[0];
+        ASSIGN_OR_RETURN(ColumnPtr new_column, _convert_to_nullable_column(columns[0], is_result_nullable, true));
+
+        // TODO: use mutable ptr as result
+        auto& ret_type = _agg_state_desc.get_return_type();
+        ColumnPtr result = ColumnHelper::create_column(ret_type, _agg_state_desc.is_result_nullable());
+        auto chunk_size = columns[0]->size();
+
+        // finalize agg states into result
+        auto align_size = _function->alignof_size();
+        auto state_size = align_to(_function->size(), align_size);
+        AggDataPtr agg_state = reinterpret_cast<AggDataPtr>(std::aligned_alloc(align_size, state_size));
+        if (UNLIKELY(agg_state == nullptr)) {
+            return Status::InternalError("Failed to allocate memory for aggregate state");
+        }
+
+        // `count` is a special case because `CountNullableAggregateFunction` is used to handle nullable column
+        // and its serialize/finalize is meant to not nullable.
+        DCHECK_EQ(is_result_nullable, new_column->is_nullable());
+        std::string function_name = _function->get_name();
+        if (function_name == "count" || function_name == "count_nullable") {
+            auto* data_column = ColumnHelper::get_data_column(new_column.get());
+            for (size_t i = 0; i < chunk_size; i++) {
+                if (new_column->is_null(i)) {
+                    result->append_nulls(1);
+                    continue;
+                }
+                _function->create(_nested_ctx, agg_state);
+                _function->merge(_nested_ctx, data_column, agg_state, i);
+                _function->finalize_to_column(_nested_ctx, agg_state, result.get());
+                _function->destroy(_nested_ctx, agg_state);
+            }
+        } else {
+            for (size_t i = 0; i < chunk_size; i++) {
+                if (new_column->is_null(i)) {
+                    result->append_nulls(1);
+                    continue;
+                }
+                _function->create(_nested_ctx, agg_state);
```

Comment:
why does it need to create & destroy for each row ?

Reply:
for `_state_union`/`_state_merge`, each row is mapping the specific group-by key's final intermeidate state,  we need output each row's intermediate/final result.